src/v/cloud_topics/level_one/metastore.h

```diff
@@ -109,6 +110,96 @@ class metastore {
     // `out_of_range`.
     virtual ss::future<std::expected<object_response, errc>>
     get_first_ge(const model::topic_id_partition&, model::timestamp) = 0;
+
+    // Compaction metadata updates per partition
+    //
+    // Kafka compaction works by taking "dirty" ranges of data, collecting the
+    // keys and offsets within that range, and then removing all older
+    // instances of those keys from the beginning of the log. At that point,
+    // that range of data is considered "cleaned". For all keys in a cleaned
+    // range, only the latest version of that key exists in the log up to the
+    // cleaned range.
+    //
+    // Kafka also has the concept of a "tombstone" (empty-value record) that
+    // indicates the logical deletion of the record's key. Once a given
+    // tombstone record has been cleaned, a timer begins and after
+    // delete.retention.ms elapses, the cleaned tombstone record may be
+    // removed.
+    //
+    // To support these ideas, the metastore tracks cleaned offset ranges and
+    // whether they contain tombstones. This allows it to expose dirty ranges
+    // and ranges with removable tombstones to callers.
+    struct compaction_update {
+        bool operator==(const compaction_update&) const = default;
+        struct cleaned_range {
+            bool operator==(const cleaned_range&) const = default;
+            kafka::offset base_offset;
+            kafka::offset last_offset;
+            std::optional<model::timestamp> cleaned_with_tombstones_at;
+        };
+        // A range indicating that the data's keys have been fully deduplicated
+        // from the start of the log.
+        std::optional<cleaned_range> new_cleaned_range;
+
+        // Ranges of offsets that previously had tombstones, that have been
+        // removed.
+        chunked_vector<cleaned_range> removed_tombstones_ranges;
+    };
+    using compaction_map_t
+      = chunked_hash_map<model::topic_id_partition, compaction_update>;
+    struct compaction_offsets_response {
+        // Offset ranges whose keys have not been fully deduplicated from the
+        // start of the log.
+        offset_interval_set dirty_ranges;
+
+        // The set of offset ranges that contain tombstones whose keys have
+        // been cleaned long enough to be eligible for tombstone removal.
+        //
+        // A compaction method, when iterating over a tombstone record, may
+        // consult this to determine if the tombstone should be removed.
+        offset_interval_set removable_tombstone_ranges;
+    };
+    // Similar to replace_objects(), but with additional constraints based on
+    // compaction metadata.
```

Comment:
maybe a note to see `get_compaction_offsets` for more information, or place this method after `get_compaction_offsets` so they have to scroll by?